keps/sig-scheduling/5075-dra-consumable-capacity/README.md

```diff
@@ -0,0 +1,1256 @@
+The number of shares can be limited if needed.
+
+### Goals
+
+- Introduce an ability to allocating on-demand shared devices via DRA. This should cover the use cases of macvlan or ipvlan in a DRA driver for CNI and virtual accelerator devices with on-demand memory fraction.
+- Enhance a capability of secondary networks to dynamically allocate secondary networks based on present availabilities such as bandwidth.
+- Leverage the enhancement of capacity consumption from [KEP-4815](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/4815-dra-partitionable-devices).
+
+### Non-Goals
+
+- Define driver-specific attributes and configs (such as CNI parameter config).
+- Support network security policy.
+- Add a flag to allow sharing device within the same pod (has no use case).
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### User Stories (Optional)
+
+#### Story 1
+A user requests a secondary network based on their bandwidth demands without wanting to specify a device name nor present available bandwidth of each specific device. The DRA scheduler can dynamically select available devices based on their availability.
+
+```yaml
+kind: ResourceSlice
+...
+spec:
+  driver: cni.dra.networking.x-k8s.io
+  devices:
+  - name: eth1
+    basic:
+      attributes:
+        name:
+          string: "eth1"
+      consumableCapacity:
+        bandwidth:
+          value: 10Gi
+---
+kind: DeviceClass
+metadata:
+  name: vlan-cni.networking.x-k8s.io
+```
+
+```yaml
+kind: ResourceClaim
+...
+spec:
+  devices:
+    requests:
+    - name: macvlan
+      deviceClassName: vlan-cni.networking.x-k8s.io
+      allocationMode: Shared
+      resources:
+        requests:
+          bandwidth: "1Gi"
+    config:
+    - requests:
+      - macvlan
+      opaque:
+        driver: cni.dra.networking.x-k8s.io
+        parameters: # CNIParameters with the GVK, interface name and CNI Config (in YAML format).
+          apiVersion: cni.networking.x-k8s.io/v1alpha1
+          kind: CNI
+          ifName: "net1"
+          config:
+            cniVersion: 1.0.0
+            name: net1
+            plugins:
+            - type: macvlan
+              master: eth0
+              mode: bridge
+              ipam:
+                type: host-local
+                ranges:
+                - - subnet: 10.10.1.0/24
+```
+
+#### Story 2
+A user reserves that fractions of memory for the virtual GPU.
+
+```yaml
+apiVersion: resource.k8s.io/v1beta1
+kind: ResourceClaim
+...
+spec:
+  devices:
+    requests:
+    - name: vgpu0
```

Comment:
Should this also set `allocationMode` to `Shared`?

Reply:
@johnbelamaric I have added `AllowShared` mode to cover a new Story 7 A user requests a device regardless of whether the device is sharable or not. 

For `AllowMultiple` and `AllowMultipleOrShared `, can we have that for a separate KEP?

Reply:
So I'm not sure I follow why sharing implies <1. Wouldn't it be possible for someone to request more than one device, but also accept that one or more of those devices might be shared? I'm not sure if it makes sense in the network use-case, but it does seem like something that might be useful for other use-cases.

I'm also uncertain about adding an additional option for the `AllocationMode` enum. I think I agree that combining `AllocationMode: All` with shared devices shouldn't be possible, it still feels a bit strange to add the new `AllowShared` value. The current values `All` and `ExactCount` specify how many devices that will be allocated for the request, which seems separate from whether or not shared devices can be used to satisfy the claim.

Reply:
> For `AllowMultiple` and `AllowMultipleOrShared `, can we have that for a separate KEP?

Yes, for sure. Although based on Morten's comment maybe we should think about other ways to specify this? Maybe `AllocationMode` isn't the right place? Do you have an alternative suggestion, @mortent ?